qt/aqt/browser/table/table.py

```diff
@@ -129,6 +129,8 @@ def select_single_card(self, card_id: CardId) -> None:
         self.clear_selection()
         if (row := self._model.get_card_row(card_id)) is not None:
             self._view.selectRow(row)
+            # editor may pop up and hide the row later on
+            QTimer.singleShot(100, lambda: self._scroll_to_row(row))
 
```

Comment:
The interface would look a bit more polished if we could open to the correct scroll location the first time. Sometimes these delay hacks are unavoidable, but I wonder if that's the case here? The following changes prevent the redraw for me - would you mind having a play with them to see if they cause any regressions?

```diff
diff --git a/qt/aqt/browser/browser.py b/qt/aqt/browser/browser.py
index 194c60def..52cdc7b07 100644
--- a/qt/aqt/browser/browser.py
+++ b/qt/aqt/browser/browser.py
@@ -106,9 +106,9 @@ class Browser(QMainWindow):
         self.setupEditor()
         # disable undo/redo
         self.on_undo_state_change(mw.undo_actions_info())
-        self.setupSearch(card, search)
         gui_hooks.browser_will_show(self)
         self.show()
+        self.setupSearch(card, search)
 
     def on_operation_did_execute(
         self, changes: OpChanges, handler: Optional[object]
diff --git a/qt/aqt/browser/table/table.py b/qt/aqt/browser/table/table.py
index 1cecd8086..67f2a0451 100644
--- a/qt/aqt/browser/table/table.py
+++ b/qt/aqt/browser/table/table.py
@@ -129,8 +129,7 @@ class Table:
         self.clear_selection()
         if (row := self._model.get_card_row(card_id)) is not None:
             self._view.selectRow(row)
-            # editor may pop up and hide the row later on
-            QTimer.singleShot(100, lambda: self._scroll_to_row(row))
+            self._scroll_to_row(row)
 
     # Reset
 
@@ -408,8 +407,7 @@ class Table:
             current = current or rows[0]
             self._select_rows(rows)
             self._set_current(current)
-            # editor may pop up and hide the row later on
-            QTimer.singleShot(100, lambda: self._scroll_to_row(current))
+            self._scroll_to_row(current)
         if self.len_selection() == 0:
             # no row change will fire
             self.browser.onRowChanged(QItemSelection(), QItemSelection())
```

Reply:
I'll merge this in for now as it's about time to get another build out, but if you spot any problems, please let me know or feel free to revert it in a follow-up PR.

Reply:
The problem the timer is supposed to solve only seems to occur when the browser is already open, so this commit doesn't help I'm afraid. To reproduce:
1. Open the browser while reviewing.
2. Scroll up until the selected row is hidden.
3. Select multiple rows to hide the editor.
4. Repeat step 1.

The editor should now conceal the selected row.

But I'm in favour of removing the timer. It can't be trusted to work in all situations, it makes debugging much harder (as I had to learn the hard way), and it gets worse exponentially if there's more than one.
I'll investigate if there's another way.

Reply:
One indirect way to "solve" it might be to change to PositionAtTop instead of PositionAtCenter :-)

Reply:
Does this work on your machine? I had tried it and it didn't make any difference. The current `PositionAtCenter` hint doesn't seem to be obeyed, either.

Reply:
This seems to work for me when applied to latest HEAD:

```difff
diff --git a/qt/aqt/browser/table/table.py b/qt/aqt/browser/table/table.py
index 437277ed2..157bdc0e6 100644
--- a/qt/aqt/browser/table/table.py
+++ b/qt/aqt/browser/table/table.py
@@ -458,7 +458,7 @@ class Table:
         visible = top_border >= 0 and bottom_border < self._view.viewport().height()
         if not visible or scroll_even_if_visible:
             horizontal = self._view.horizontalScrollBar().value()
-            self._view.scrollTo(self._model.index(row, 0), self._view.PositionAtCenter)
+            self._view.scrollTo(self._model.index(row, 0), self._view.PositionAtTop)
             self._view.horizontalScrollBar().setValue(horizontal)
 
     def _scroll_to_column(self, column: int) -> None:
``` 

Following the 4 steps you listed above, it always scrolls to the top

Reply:
Must be a Windows issue then. Ugh.

Reply:
Works for me on Windows - maybe I'm doing something different? Before and after pressing b in the main window:

<img width="1506" alt="Screen Shot 2021-06-09 at 8 31 51 am" src="https://user-images.githubusercontent.com/180542/121266116-50a1e380-c8fd-11eb-9e16-65f7490d68a6.png">
<img width="1500" alt="Screen Shot 2021-06-09 at 8 31 59 am" src="https://user-images.githubusercontent.com/180542/121266108-4da6f300-c8fd-11eb-9eb7-32bca38f6bbd.png">

Reply:
Looks like what I'm doing. Here's a GIF:

![3qTRB0GWrO](https://user-images.githubusercontent.com/48286681/121299429-b4c0a980-c8f5-11eb-8db2-b8567e409c5e.gif)

As always, no add-ons involved. This is with `PositionAtCenter`, but as can be seen, it scrolls until the row appears at the bottom. If I had scrolled down on the other hand, it would scroll until the row appeared at the top. So it appears that the behaviour is always `EnsureVisible`, regardless of the actual scrollHint set.

Reply:
Okay, I know what happened. `selectRow()` already scrolls the row into view, then `_scroll_to_row()` isn't executed because the row is already visible (before the editor pops up). I hadn't pulled c5e56a5fe81d1c56ea1c415d086cee96c7f75b8b which fixes the issue by scrolling independently of visibility.
Another fix would be to scroll before selecting:

```diff
diff --git a/qt/aqt/browser/table/table.py b/qt/aqt/browser/table/table.py
index 67f2a045..f9fb9e4c 100644
--- a/qt/aqt/browser/table/table.py
+++ b/qt/aqt/browser/table/table.py
@@ -128,8 +128,8 @@ class Table:
         """Try to set the selection to the item corresponding to the given card."""
         self.clear_selection()
         if (row := self._model.get_card_row(card_id)) is not None:
-            self._view.selectRow(row)
             self._scroll_to_row(row)
+            self._view.selectRow(row)

     # Reset
```

However, the issue remains that we scroll before the editor shows. So the row isn't centered in the reduced view, or may even be hidden if the editor covers more than half of the vertical space. But like you said in the beginning, this won't be a problem if we `PositionAtTop` as in your screenshots.